Add concurrency option

dudemcbacon opened the pull request:
Adding concurrency option to control the number of concurrent hosts that SSH is run on.

mmoll commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -1,6 +1,7 @@
 require 'hammer_cli'
 require 'hammer_cli_foreman/host'
 require 'net/ssh/multi'
+require 'pry'
```
leftover from testing? :)

dudemcbacon replied:
Derp. Let me get rid of that...

dudemcbacon commented:
I also noticed there were some race conditions affecting net-ssh-multi -- fortunately, they have been fixed in an upstream pull request. I've updated the gemspec to reflect the new dependency.

See [this pull request](https://github.com/net-ssh/net-ssh-multi/pull/4) for more information on the race conditions.

mmoll commented:
Thats, OK, we can handle the version dependency in our packaging, I guess...
@ohadlevy, could you test?

mmoll commented:
@dudemcbacon BTW, could you squash your commits together into one?

dudemcbacon commented:
Done!

ohadlevy commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -12,6 +12,7 @@ class Command < HammerCLIForeman::Command
 
     command_name 'SSH to hosts'
     option %w(-c --command), 'COMMAND', _('Command to execute'), :attribute_name => :command, :required => true
+    option %w(-n --concurrent), 'CONCURRENCY', _('Number of concurrent SSH sessions'), :attribute_name => :concurrent
```
should we provide a default? i think asking each time could be annoying ?

dudemcbacon replied:
I'd actually prefer if we didn't. It'd make the code more complex. When the option isn't set and no default is specified concurrent == nil. When concurrent_connections is set to nil it disables the concurrency limit. If we set a default value for concurrent it will be set to whatever the default is even if the option isn't specified. We'd either need to add an option to disable concurrency limits or use some other method to determine when the option wasn't actually passed by the user.

mmoll replied:
"it disables the concurrency limit." means 1 or infinity?

dudemcbacon replied:
infinity -- as in the way it behaved before this pull request.

mmoll replied:
I guess if it's like before I'm OK with it - and see this PR even more valuable as it's making a limit possible. :dart:

mbacovsky replied:
You can add
```ruby
option ....... do |o|
  Integer(o)
end
```
for Hammer to ensure valid Integer is given

+1 to not force user to set concurrency limit

dudemcbacon commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -36,10 +37,16 @@ def execute
       ssh_options[:keys] = [identity_file] unless identity_file.to_s.empty?
 
-      Net::SSH::Multi.start(:on_error => :warn) do |session|
+      Net::SSH::Multi.start(:concurrent_connections => concurrent.to_i, :on_error => :warn) do |session|
```
It looks like the assumption I made here was that concurrent.to_i would return nil if concurrent was set to nil. It actually returns 0. This makes my statements in regards to line 15 somewhat non-sensical. I will update the pull request to have the behavior I intended.

dudemcbacon commented:
I have modified the pull request slightly in the following ways:

1. move the unless logic for concurrent to the top of the execute method -- that way if the parameter supplied for concurrent is 0 it will display an error message before the host list and confirmation prompt.

2. Added a variable 'limit' which is set to either nil or an integer depending on whether the user sets the concurrent option when they run hammer.

3. Concurrent_connections is set to the limit variable (as either nil or an integer > 0).

This fixes the assumption I made about the .to_i method when used with nil and assures the following behavior:

1. If the concurrent option is specified with an integer > 0 then the number of concurrent connections Net::SSH::Multi will allow is limited to that number.

2. If the concurrent option is not specified than the original behavior of not limiting the number of concurrent_connections will be present.

mbacovsky commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -28,6 +29,13 @@ def request_params
     def execute
+      if concurrent && concurrent.to_i == 0
+        warn _("specify 1 or more concurrent hosts")
+        return HammerCLI::EX_OK
```
I'd prefer to 
```ruby
signal_usage_error(_("specify 1 or more concurrent hosts")) if (!concurrent.nil? && concurrent < 1)
```
to get proper exit code and error handling

mbacovsky commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -28,6 +29,13 @@ def request_params
+        return HammerCLI::EX_OK
+      end
+
+      limit = concurrent.nil? ? concurrent : concurrent.to_i
```
Should be ensured in the option block

mbacovsky commented on lib/hammer_cli_foreman/ssh.rb:
```diff
@@ -39,7 +47,8 @@ def execute
 
-      Net::SSH::Multi.start(:on_error => :warn) do |session|
+      Net::SSH::Multi.start(:concurrent_connections => limit, :on_error => :warn) do |session|
+        puts _("executing on #{limit} concurrent host(s)") if limit.to_i > 0
```
I'm not aware about how important this information is for the user, but FYI there is also option to just log it with `logger.info(_(.....))`

mbacovsky commented:
Besides the minor things mentioned inline it looks good. Thanks @dudemcbacon!

mbacovsky commented:
@dudemcbacon, could you please rebase?

dudemcbacon commented:
I have rebased this pull request and modified the code as suggested by @mbacovsky.

mbacovsky commented on lib/hammer_cli_foreman_ssh/ssh.rb:
```diff
@@ -12,6 +12,10 @@ class Command < HammerCLIForeman::Command
     option %w(-c --command), 'COMMAND', _('Command to execute'), :attribute_name => :command, :required => true
+    option %w(-n --concurrent), 'CONCURRENCY', _('Number of concurrent SSH sessions'), :attribute_name => :concurrent do |o|
+      Integer(o)
+      o.nil? ? o : o.to_i
```
this line is not neccessary, as the block is executed only if the parameter was set

dudemcbacon replied:
Ahh, good catch. I guess I didn't think about it that much. I will remove it.

mbacovsky commented:
@dudemcbacon, thanks for the update! Looks good now!
I'm curious about the extra line in the option block, was it added to any unexpected behaviour?

dudemcbacon commented:
I have removed the line you commented on in the option block.

dudemcbacon commented:
The line you're referring to was to ensure that the concurrent variable was set to either nil or a number > 0. As you said in your inline comment, it's not necessary. If the option block is called it will be a number. If the option block is not called, it will be nil. If the integer is 0, it will be filtered by the signal_usage_error method.

mbacovsky commented:
:+1:  @dudemcbacon, that makes sense. Thank you for your patience ;) :beer:  Merging...

dudemcbacon commented:
w00t!
